File: packages/sdk/src/modules/batch.ts
/**
 * ProtocolBanks SDK - Batch Payment Module
 * 
 * 批量支付，支持:
 * - 批量验证 (最多 500 收款人)
 * - 批量提交
 * - 状态追踪
 * - 失败重试
 */

import type {
  BatchRecipient,
  BatchValidationError,
  BatchSubmitResult,
  BatchStatus,
  BatchItemStatus,
  BatchOptions,
  ChainId,
  TokenSymbol,
  IBatchModule,
} from '../types';
import { ErrorCodes } from '../types';
import { ProtocolBanksError } from '../utils/errors';
import { generateUUID } from '../utils/crypto';
import {
  isValidAddress,
  isValidAmount,
  isValidToken,
  validateBatchSize,
  detectHomoglyphs,
} from '../utils/validation';
import { MAX_BATCH_SIZE, getTokenDecimals, parseAmount } from '../config/chains';
import type { HttpClient } from '../utils/http';

// ============================================================================
// Batch Module Implementation
// ============================================================================

export class BatchModule implements IBatchModule {
  private http: HttpClient;
  private batches: Map<string, BatchStatus> = new Map();
  private pollingIntervals: Map<string, NodeJS.Timeout> = new Map();
  
  constructor(http: HttpClient) {
    this.http = http;
  }
  
  // ============================================================================
  // Public Methods
  // ============================================================================
  
  /**
   * Validate batch recipients
   * Returns errors for ALL invalid entries (not just first)
   */
  async validate(recipients: BatchRecipient[]): Promise<BatchValidationError[]> {
    const errors: BatchValidationError[] = [];
    
    // Check batch size
    if (recipients.length > MAX_BATCH_SIZE) {
      throw new ProtocolBanksError({
        code: ErrorCodes.BATCH_SIZE_EXCEEDED,
        category: 'BATCH',
        message: `Batch size ${recipients.length} exceeds maximum of ${MAX_BATCH_SIZE}`,
        details: { size: recipients.length, maxSize: MAX_BATCH_SIZE },
        retryable: false,
      });
    }
    
    if (recipients.length === 0) {
      throw new ProtocolBanksError({
        code: ErrorCodes.BATCH_VALIDATION_FAILED,
        category: 'BATCH',
        message: 'Batch cannot be empty',
        retryable: false,
      });
    }
    
    // Validate each recipient
    for (let i = 0; i < recipients.length; i++) {
      const recipient = recipients[i]!;
      const recipientErrors: string[] = [];
      
      // Validate address
      if (!recipient.address) {
        recipientErrors.push('Address is required');
      } else {
        // Check for homoglyphs
        const homoglyphDetails = detectHomoglyphs(recipient.address);
        if (homoglyphDetails) {
          recipientErrors.push('Address contains suspicious characters (possible homoglyph attack)');
        } else if (!isValidAddress(recipient.address)) {
          recipientErrors.push('Invalid address format');
        }
      }
      
      // Validate amount
      if (!recipient.amount) {
        recipientErrors.push('Amount is required');
      } else if (!isValidAmount(recipient.amount)) {
        recipientErrors.push('Invalid amount (must be positive, max 1 billion)');
      }
      
      // Validate token
      if (!recipient.token) {
        recipientErrors.push('Token is required');
      } else if (!isValidToken(recipient.token)) {
        recipientErrors.push(`Unsupported token: ${recipient.token}`);
      }
      
      // Validate memo length
      if (recipient.memo && recipient.memo.length > 256) {
        recipientErrors.push('Memo exceeds maximum length of 256 characters');
      }
      
      // Add errors if any
      if (recipientErrors.length > 0) {
        errors.push({
          index: i,
          address: recipient.address || '',
          errors: recipientErrors,
        });
      }
    }
    
    // Check for duplicate addresses
    const addressCounts = new Map<string, number[]>();
    for (let i = 0; i < recipients.length; i++) {
      const address = recipients[i]!.address?.toLowerCase();
      if (address) {
        const indices = addressCounts.get(address) ?? [];
        indices.push(i);
        addressCounts.set(address, indices);
      }
    }
    
    for (const [address, indices] of addressCounts) {
      if (indices.length > 1) {
        // Add warning for duplicates (not error, just warning)
        for (const index of indices) {
          const existing = errors.find(e => e.index === index);
          if (existing) {
            existing.errors.push(`Duplicate address (appears ${indices.length} times)`);
          } else {
            errors.push({
              index,
              address,
              errors: [`Duplicate address (appears ${indices.length} times)`],
            });
          }
        }
      }
    }
    
    return errors.sort((a, b) => a.index - b.index);
  }
  
  /**
   * Submit batch payment
   */
  async submit(
    recipients: BatchRecipient[],
    options?: BatchOptions
  ): Promise<BatchSubmitResult> {
    // Validate batch
    const validationErrors = await this.validate(recipients);
    
    // Filter out warnings (duplicates) vs errors
    const criticalErrors = validationErrors.filter(e => 
      !e.errors.every(err => err.includes('Duplicate'))
    );
    
    if (criticalErrors.length > 0) {
      return {
        batchId: '',
        status: 'failed',
        validCount: recipients.length - criticalErrors.length,
        invalidCount: criticalErrors.length,
        errors: criticalErrors,
      };
    }
    
    // Generate batch ID
    const batchId = `batch_${generateUUID().replace(/-/g, '')}`;
    
    // Calculate total amount
    const totalAmount = recipients.reduce((sum, r) => {
      return sum + parseFloat(r.amount);
    }, 0);
    
    // Prepare batch items
    const items: BatchItemStatus[] = recipients.map((r, index) => ({
      index,
      address: r.address,
      amount: r.amount,
      status: 'pending' as const,
    }));
    
    // Create batch status
    const batchStatus: BatchStatus = {
      batchId,
      status: 'pending',
      progress: {
        total: recipients.length,
        completed: 0,
        failed: 0,
        pending: recipients.length,
      },
      items,
      totalAmount: totalAmount.toString(),
      createdAt: new Date(),
    };
    
    // Store locally
    this.batches.set(batchId, batchStatus);
    
    // Submit to backend
    try {
      const response = await this.http.post<BatchSubmitResult>('/batch/submit', {
        batchId,
        recipients: recipients.map(r => ({
          address: r.address,
          amount: r.amount,
          token: r.token,
          memo: r.memo,
          orderId: r.orderId,
        })),
        chain: options?.chain,
        priority: options?.priority ?? 'medium',
        webhookUrl: options?.webhookUrl,
        idempotencyKey: options?.idempotencyKey,
      });
      
      // Update local status
      batchStatus.status = response.status;
      
      return {
        batchId,
        status: response.status,
        validCount: recipients.length,
        invalidCount: 0,
        errors: [],
        estimatedFee: response.estimatedFee,
      };
      
    } catch (error) {
      batchStatus.status = 'failed';
      throw error;
    }
  }
  
  /**
   * Get batch status
   */
  async getStatus(batchId: string): Promise<BatchStatus> {
    // Check local cache first
    const localBatch = this.batches.get(batchId);
    
    // Fetch from backend
    try {
      const response = await this.http.get<BatchStatus>(`/batch/${batchId}`);
      
      // Update local cache
      if (localBatch) {
        Object.assign(localBatch, response);
      } else {
        this.batches.set(batchId, response);
      }
      
      return response;
      
    } catch (error) {
      // Return local state if available
      if (localBatch) {
        return localBatch;
      }
      
      throw new ProtocolBanksError({
        code: ErrorCodes.BATCH_NOT_FOUND,
        category: 'BATCH',
        message: `Batch ${batchId} not found`,
        retryable: false,
      });
    }
  }
  
  /**
   * Retry failed items in batch
   */
  async retry(batchId: string, itemIndices?: number[]): Promise<BatchSubmitResult> {
    const batch = await this.getStatus(batchId);
    
    // Check if batch can be retried
    if (batch.status === 'processing') {
      throw new ProtocolBanksError({
        code: ErrorCodes.BATCH_ALREADY_PROCESSING,
        category: 'BATCH',
        message: 'Batch is currently processing, cannot retry',
        retryable: false,
      });
    }
    
    // Get failed items
    const failedItems = batch.items.filter(item => 
      item.status === 'failed' && 
      (itemIndices === undefined || itemIndices.includes(item.index))
    );
    
    if (failedItems.length === 0) {
      return {
        batchId,
        status: batch.status as 'pending' | 'processing' | 'completed' | 'failed',
        validCount: 0,
        invalidCount: 0,
        errors: [],
      };
    }
    
    // Submit retry request
    const response = await this.http.post<BatchSubmitResult>(`/batch/${batchId}/retry`, {
      itemIndices: failedItems.map(i => i.index),
    });
    
    return response;
  }
  
  /**
   * Poll batch status with callback
   */
  poll(
    batchId: string,
    callback: (status: BatchStatus) => void,
    interval: number = 5000
  ): () => void {
    // Clear existing polling for this batch
    const existingInterval = this.pollingIntervals.get(batchId);
    if (existingInterval) {
      clearInterval(existingInterval);
    }
    
    // Start polling
    const pollFn = async () => {
      try {
        const status = await this.getStatus(batchId);
        callback(status);
        
        // Stop polling if batch is complete
        if (this.isFinalStatus(status.status)) {
          this.stopPolling(batchId);
        }
      } catch (error) {
        // Continue polling on error
        console.warn(`Batch polling error for ${batchId}:`, error);
      }
    };
    
    // Initial poll
    pollFn();
    
    // Set up interval
    const intervalId = setInterval(pollFn, interval);
    this.pollingIntervals.set(batchId, intervalId);
    
    // Return stop function
    return () => this.stopPolling(batchId);
  }
  
  /**
   * Stop polling for a batch
   */
  stopPolling(batchId: string): void {
    const intervalId = this.pollingIntervals.get(batchId);
    if (intervalId) {
      clearInterval(intervalId);
      this.pollingIntervals.delete(batchId);
    }
  }
  
  /**
   * Stop all polling
   */
  stopAllPolling(): void {
    for (const [batchId] of this.pollingIntervals) {
      this.stopPolling(batchId);
    }
  }
  
  // ============================================================================
  // Helper Methods
  // ============================================================================
  
  /**
   * Calculate total amount for batch
   */
  calculateTotal(recipients: BatchRecipient[]): {
    byToken: Record<string, string>;
    totalUSD: string;
  } {
    const byToken: Record<string, number> = {};
    
    for (const recipient of recipients) {
      const amount = parseFloat(recipient.amount);
      if (!isNaN(amount)) {
        byToken[recipient.token] = (byToken[recipient.token] ?? 0) + amount;
      }
    }
    
    // Convert to string
    const byTokenStr: Record<string, string> = {};
    for (const [token, amount] of Object.entries(byToken)) {
      byTokenStr[token] = amount.toFixed(6);
    }
    
    // Estimate USD total (assuming stablecoins = $1)
    const stablecoins = ['USDC', 'USDT', 'DAI'];
    let totalUSD = 0;
    for (const [token, amount] of Object.entries(byToken)) {
      if (stablecoins.includes(token)) {
        totalUSD += amount;
      }
    }
    
    return {
      byToken: byTokenStr,
      totalUSD: totalUSD.toFixed(2),
    };
  }
  
  /**
   * Get batch progress percentage
   */
  getProgress(status: BatchStatus): number {
    if (status.progress.total === 0) return 0;
    return Math.round(
      ((status.progress.completed + status.progress.failed) / status.progress.total) * 100
    );
  }
  
  // ============================================================================
  // Private Methods
  // ============================================================================
  
  private isFinalStatus(status: string): boolean {
    return ['completed', 'failed'].includes(status);
  }
}

// ============================================================================
// Factory Function
// ============================================================================

/**
 * Create a new BatchModule instance
 */
export function createBatchModule(http: HttpClient): BatchModule {
  return new BatchModule(http);
}

File: packages/python/src/protocolbanks/modules/batch.py
"""
ProtocolBanks SDK - Batch Payment Module

批量支付，支持:
- 批量验证 (最多 500 收款人)
- 批量提交
- 状态追踪
- 失败重试
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Callable

from protocolbanks.config import MAX_BATCH_SIZE
from protocolbanks.errors import ProtocolBanksError
from protocolbanks.types import (
    BatchItemStatus,
    BatchOptions,
    BatchProgress,
    BatchRecipient,
    BatchStatus,
    BatchSubmitResult,
    BatchValidationError,
    ErrorCodes,
)
from protocolbanks.utils.crypto import generate_uuid
from protocolbanks.utils.validation import (
    detect_homoglyphs,
    is_valid_address,
    is_valid_amount,
    is_valid_token,
)


class BatchModule:
    """Batch Payment Module"""

    def __init__(self, http_client: Any):
        self._http = http_client
        self._batches: dict[str, BatchStatus] = {}
        self._polling_tasks: dict[str, bool] = {}

    # ============================================================================
    # Public Methods
    # ============================================================================

    async def validate(
        self, recipients: list[BatchRecipient] | list[dict]
    ) -> list[BatchValidationError]:
        """
        Validate batch recipients.
        Returns errors for ALL invalid entries (not just first).
        """
        # Convert dicts to BatchRecipient if needed
        recipients_list: list[BatchRecipient] = [
            BatchRecipient(**r) if isinstance(r, dict) else r for r in recipients
        ]

        errors: list[BatchValidationError] = []

        # Check batch size
        if len(recipients_list) > MAX_BATCH_SIZE:
            raise ProtocolBanksError(
                code=ErrorCodes.BATCH_SIZE_EXCEEDED,
                message=f"Batch size {len(recipients_list)} exceeds maximum of {MAX_BATCH_SIZE}",
                details={"size": len(recipients_list), "max_size": MAX_BATCH_SIZE},
                retryable=False,
            )

        if len(recipients_list) == 0:
            raise ProtocolBanksError(
                code=ErrorCodes.BATCH_VALIDATION_FAILED,
                message="Batch cannot be empty",
                retryable=False,
            )

        # Validate each recipient
        for i, recipient in enumerate(recipients_list):
            recipient_errors: list[str] = []

            # Validate address
            if not recipient.address:
                recipient_errors.append("Address is required")
            else:
                # Check for homoglyphs
                homoglyph_details = detect_homoglyphs(recipient.address)
                if homoglyph_details:
                    recipient_errors.append(
                        "Address contains suspicious characters (possible homoglyph attack)"
                    )
                elif not is_valid_address(recipient.address):
                    recipient_errors.append("Invalid address format")

            # Validate amount
            if not recipient.amount:
                recipient_errors.append("Amount is required")
            elif not is_valid_amount(recipient.amount):
                recipient_errors.append("Invalid amount (must be positive, max 1 billion)")

            # Validate token
            if not recipient.token:
                recipient_errors.append("Token is required")
            elif not is_valid_token(recipient.token):
                recipient_errors.append(f"Unsupported token: {recipient.token}")

            # Validate memo length
            if recipient.memo and len(recipient.memo) > 256:
                recipient_errors.append("Memo exceeds maximum length of 256 characters")

            # Add errors if any
            if recipient_errors:
                errors.append(
                    BatchValidationError(
                        index=i,
                        address=recipient.address or "",
                        errors=recipient_errors,
                    )
                )

        # Check for duplicate addresses
        address_counts: dict[str, list[int]] = {}
        for i, recipient in enumerate(recipients_list):
            if recipient.address:
                address = recipient.address.lower()
                if address not in address_counts:
                    address_counts[address] = []
                address_counts[address].append(i)

        for address, indices in address_counts.items():
            if len(indices) > 1:
                # Add warning for duplicates
                for index in indices:
                    existing = next((e for e in errors if e.index == index), None)
                    if existing:
                        existing.errors.append(
                            f"Duplicate address (appears {len(indices)} times)"
                        )
                    else:
                        errors.append(
                            BatchValidationError(
                                index=index,
                                address=address,
                                errors=[f"Duplicate address (appears {len(indices)} times)"],
                            )
                        )

        return sorted(errors, key=lambda e: e.index)

    async def submit(
        self,
        recipients: list[BatchRecipient] | list[dict],
        options: BatchOptions | dict | None = None,
    ) -> BatchSubmitResult:
        """Submit batch payment"""
        # Convert dicts if needed
        recipients_list: list[BatchRecipient] = [
            BatchRecipient(**r) if isinstance(r, dict) else r for r in recipients
        ]

        if options and isinstance(options, dict):
            options = BatchOptions(**options)

        # Validate batch
        validation_errors = await self.validate(recipients_list)

        # Filter out warnings (duplicates) vs errors
        critical_errors = [
            e for e in validation_errors if not all("Duplicate" in err for err in e.errors)
        ]

        if critical_errors:
            return BatchSubmitResult(
                batch_id="",
                status="failed",
                valid_count=len(recipients_list) - len(critical_errors),
                invalid_count=len(critical_errors),
                errors=critical_errors,
            )

        # Generate batch ID
        batch_id = f"batch_{generate_uuid().replace('-', '')}"

        # Calculate total amount
        total_amount = sum(float(r.amount) for r in recipients_list)

        # Prepare batch items
        items: list[BatchItemStatus] = [
            BatchItemStatus(
                index=i,
                address=r.address,
                amount=r.amount,
                status="pending",
            )
            for i, r in enumerate(recipients_list)
        ]

        # Create batch status
        batch_status = BatchStatus(
            batch_id=batch_id,
            status="pending",
            progress=BatchProgress(
                total=len(recipients_list),
                completed=0,
                failed=0,
                pending=len(recipients_list),
            ),
            items=items,
            total_amount=str(total_amount),
            created_at=datetime.now(),
        )

        # Store locally
        self._batches[batch_id] = batch_status

        # Submit to backend
        try:
            response = await self._http.post(
                "/batch/submit",
                {
                    "batch_id": batch_id,
                    "recipients": [
                        {
                            "address": r.address,
                            "amount": r.amount,
                            "token": r.token,
                            "memo": r.memo,
                            "order_id": r.order_id,
                        }
                        for r in recipients_list
                    ],
                    "chain": options.chain if options else None,
                    "priority": options.priority if options else "medium",
                    "webhook_url": options.webhook_url if options else None,
                    "idempotency_key": options.idempotency_key if options else None,
                },
            )

            # Update local status
            batch_status.status = response.get("status", "pending")

            return BatchSubmitResult(
                batch_id=batch_id,
                status=response.get("status", "pending"),
                valid_count=len(recipients_list),
                invalid_count=0,
                errors=[],
                estimated_fee=response.get("estimated_fee"),
            )

        except Exception as e:
            batch_status.status = "failed"
            raise e

    async def get_status(self, batch_id: str) -> BatchStatus:
        """Get batch status"""
        # Check local cache first
        local_batch = self._batches.get(batch_id)

        # Fetch from backend
        try:
            response = await self._http.get(f"/batch/{batch_id}")

            # Update local cache
            if local_batch:
                local_batch.status = response.get("status", local_batch.status)
                local_batch.progress = BatchProgress(**response.get("progress", {}))
                local_batch.items = [
                    BatchItemStatus(**item) for item in response.get("items", [])
                ]
            else:
                batch_status = BatchStatus(**response)
                self._batches[batch_id] = batch_status
                return batch_status

            return local_batch

        except Exception:
            # Return local state if available
            if local_batch:
                return local_batch

            raise ProtocolBanksError(
                code=ErrorCodes.BATCH_NOT_FOUND,
                message=f"Batch {batch_id} not found",
                retryable=False,
            )

    async def retry(
        self, batch_id: str, item_indices: list[int] | None = None
    ) -> BatchSubmitResult:
        """Retry failed items in batch"""
        batch = await self.get_status(batch_id)

        # Check if batch can be retried
        if batch.status == "processing":
            raise ProtocolBanksError(
                code=ErrorCodes.BATCH_ALREADY_PROCESSING,
                message="Batch is currently processing, cannot retry",
                retryable=False,
            )

        # Get failed items
        failed_items = [
            item
            for item in batch.items
            if item.status == "failed"
            and (item_indices is None or item.index in item_indices)
        ]

        if not failed_items:
            return BatchSubmitResult(
                batch_id=batch_id,
                status=batch.status,  # type: ignore
                valid_count=0,
                invalid_count=0,
                errors=[],
            )

        # Submit retry request
        response = await self._http.post(
            f"/batch/{batch_id}/retry",
            {"item_indices": [i.index for i in failed_items]},
        )

        return BatchSubmitResult(**response)

    def poll(
        self,
        batch_id: str,
        callback: Callable[[BatchStatus], None],
        interval: float = 5.0,
    ) -> Callable[[], None]:
        """Poll batch status with callback"""
        import asyncio

        self._polling_tasks[batch_id] = True

        async def poll_fn():
            while self._polling_tasks.get(batch_id, False):
                try:
                    status = await self.get_status(batch_id)
                    callback(status)

                    # Stop polling if batch is complete
                    if self._is_final_status(status.status):
                        self.stop_polling(batch_id)
                        break

                except Exception:
                    # Continue polling on error
                    pass

                await asyncio.sleep(interval)

        # Start polling in background
        asyncio.create_task(poll_fn())

        # Return stop function
        return lambda: self.stop_polling(batch_id)

    def stop_polling(self, batch_id: str) -> None:
        """Stop polling for a batch"""
        self._polling_tasks[batch_id] = False

    def stop_all_polling(self) -> None:
        """Stop all polling"""
        for batch_id in list(self._polling_tasks.keys()):
            self.stop_polling(batch_id)

    # ============================================================================
    # Helper Methods
    # ============================================================================

    def calculate_total(
        self, recipients: list[BatchRecipient]
    ) -> dict[str, Any]:
        """Calculate total amount for batch"""
        by_token: dict[str, float] = {}

        for recipient in recipients:
            try:
                amount = float(recipient.amount)
                by_token[recipient.token] = by_token.get(recipient.token, 0) + amount
            except (ValueError, TypeError):
                pass

        # Convert to string
        by_token_str = {token: f"{amount:.6f}" for token, amount in by_token.items()}

        # Estimate USD total (assuming stablecoins = $1)
        stablecoins = ["USDC", "USDT", "DAI"]
        total_usd = sum(
            amount for token, amount in by_token.items() if token in stablecoins
        )

        return {
            "by_token": by_token_str,
            "total_usd": f"{total_usd:.2f}",
        }

    def get_progress(self, status: BatchStatus) -> int:
        """Get batch progress percentage"""
        if status.progress.total == 0:
            return 0
        return round(
            ((status.progress.completed + status.progress.failed) / status.progress.total)
            * 100
        )

    # ============================================================================
    # Private Methods
    # ============================================================================

    def _is_final_status(self, status: str) -> bool:
        """Check if status is final"""
        return status in ("completed", "failed")


def create_batch_module(http_client: Any) -> BatchModule:
    """Create a new BatchModule instance"""
    return BatchModule(http_client)
